### ISSUE TITLE
Temporary files are not deleted

### ISSUE THREAD

**Issue by jarro2783** [logstash-plugins/logstash-output-s3#122]
It looks like temporary files used for uploading are not being deleted. Over time the number of temporary files in `/tmp/logstash` is growing (I've seen 8000 files in one day). I see messages like this
```
{"level":"DEBUG","loggerName":"logstash.outputs.s3","timeMillis":1484697405218,"thread":"S3 output uploader, file: /tmp/logstash/226d7b5d-ff1d-4486-8eb8-5571ebdd7386/ap-southeast-2/2016-10-24/telegraf/ls.s3.5f5af9cb-6e69-42ca-bb44-3a0b7ea31973.2017-01-17T18.44.part1.txt","logEvent":{"message":"Removing temporary file","file":"/tmp/logstash/226d7b5d-ff1d-4486-8eb8-5571ebdd7386/ap-southeast-2/2016-10-24/telegraf/ls.s3.5f5af9cb-6e69-42ca-bb44-3a0b7ea31973.2017-01-17T18.44.part1.txt"}}
```
but that file still exists.

I think this is the actual cause of what we are seeing in #120. @joshuaspence.

Checking the bucket, I think files are being uploaded correctly, and the uploads are keeping up.

- Version: Logstash 5.1.1, logstash-output-s3 4.0.2
- Operating System: Ubuntu 16.04
- Config File
```
output {
  s3 {
    bucket => "__redacted__"
    codec => "json_lines"
    prefix => "ap-southeast-2/%{+yyyy-MM-dd}/%{type}/"
    region => "ap-southeast-2"
    size_file => 104857600
    time_file => 5
    validate_credentials_on_root_bucket => false
  }
}
```
- Steps to Reproduce: Run logstash for longer than the `time_file` rotation period.

**Comment by jcsorvasi** [comment]
I can second this.

logstash version: docker 5.1.1
s3 output plugin version: 4.0.3
plugin config:

```
output {
  s3 {
    bucket => "some_bucket"
    access_key_id => "some_key"
    secret_access_key => "some_secret"
    region => "eu-west-1"
    codec => "json_lines"
    encoding => "none"
    size_file => "16777216"
    time_file => "1"
    upload_workers_count => "8"
    restore => "true"
  }
}
```

When trying to migrate from an old combination of logstash 2.4.1 + s3 output plugin 3.2.0 to the above described latest versions, the output plugin now seems to fail to clean up after itself. The temporary directories are created inside /tmp/logstash with a file inside them like so:

```
/tmp/logstash/02880db8-7da0-40e3-a64a-e7899d06153f/ls.s3.49146331-8fc2-48a7-8ef8-5a9854f444fc.2017-01-12T12.23.part27.txt
```

The files are uploaded to s3, but they remain on disk eventually filling it up.
An excerpt from the logs:

```
$ cat stdout | grep c79e45ffdc7.2017-01-12T12.41.part59
12:41:50.190 [[main]>worker1] DEBUG logstash.outputs.s3 - Rotate file {:strategy=>"LogStash::Outputs::S3::SizeAndTimeRotationPolicy", :key=>"ls.s3.ca3c5216-d807-4386-8c40-fc79e45ffdc7.2017-01-12T12.41.part59.txt", :path=>"/tmp/logstash/30818e0b-a354-4ab0-be56-8b8e9c29645c/ls.s3.ca3c5216-d807-4386-8c40-fc79e45ffdc7.2017-01-12T12.41.part59.txt"}
12:41:50.190 [[main]>worker1] DEBUG logstash.outputs.s3 - Queue for upload {:path=>"/tmp/logstash/30818e0b-a354-4ab0-be56-8b8e9c29645c/ls.s3.ca3c5216-d807-4386-8c40-fc79e45ffdc7.2017-01-12T12.41.part59.txt"}
12:41:53.538 [S3 output uploader, file: /tmp/logstash/30818e0b-a354-4ab0-be56-8b8e9c29645c/ls.s3.ca3c5216-d807-4386-8c40-fc79e45ffdc7.2017-01-12T12.41.part59.txt] DEBUG logstash.outputs.s3 - Removing temporary file {:file=>"/tmp/logstash/30818e0b-a354-4ab0-be56-8b8e9c29645c/ls.s3.ca3c5216-d807-4386-8c40-fc79e45ffdc7.2017-01-12T12.41.part59.txt"}
```

Yet the directory and the file inside it remain on disk.

Might be worth to add, that logstash is running in docker with the volume `/tmp/logstash` is mapped out onto the host's `/tmp/logstash`. The same configuration, same setup didn't have this problem with the previous versions.

**Comment by jarro2783** [comment]
We don't have docker involved in our case.

**Comment by ph** [comment]
I can confirm this bug with adding new expectation in one of the integration test that check for stales writes.

```
File Time rotation with stale write
  create one file
  Persists all events (FAILED - 1)

Failures:

  1) File Time rotation with stale write Persists all events
     Failure/Error: expect(Dir.glob(File.join(temporary_directory, "**", "*.txt")).size).to eq(0)

       expected: 0
            got: 2

       (compared using ==)
     # ./spec/integration/time_based_rotation_with_stale_write_spec.rb:60:in `(root)'
     # ./vendor/jruby/1.9/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `(root)'

Finished in 13.63 seconds (files took 7.56 seconds to load)
2 examples, 1 failure

Failed examples:

rspec ./spec/integration/time_based_rotation_with_stale_write_spec.rb:43 # File Time rotation with stale write Persists all events
```

**Comment by ph** [comment]
I've made a fix in #123, I will try to get that merge asap.
